toolchain/check/handle_where.cpp

```diff
@@ -17,44 +21,79 @@ auto HandleParseNode(Context& context, Parse::WhereOperandId node_id) -> bool {
+  if (auto facet_type = context.types().TryGetAs<SemIR::FacetType>(
+          self_without_constraints_type_id)) {
+    const auto& info = context.facet_types().Get(facet_type->facet_type_id);
+    auto stripped_info =
+        SemIR::FacetTypeInfo{.extend_constraints = info.extend_constraints};
+    stripped_info.Canonicalize();
+    self_without_constraints_type_id = GetFacetType(context, stripped_info);
   }
 
   // Introduce a name scope so that we can remove the `.Self` entry we are
   // adding to name lookup at the end of the `where` expression.
   context.scope_stack().PushForSameRegion();
   // Introduce `.Self` as a symbolic binding. Its type is the value of the
   // expression to the left of `where`, so `MyInterface` in the example above.
+  //
+  // It uses the `Self` type _without_ any constraints so that the `Self` is the
+  // same at all levels of a nested facet type such as:
+  //   `(Z where .X = {}) where .X = {}`.
   auto entity_name_id = context.entity_names().AddCanonical(
       {.name_id = SemIR::NameId::PeriodSelf,
        .parent_scope_id = context.scope_stack().PeekNameScopeId()});
-  auto inst_id = AddInst(
+  auto period_self_inst_id = AddInst(
       context, SemIR::LocIdAndInst::NoLoc<SemIR::BindSymbolicName>({
-                   .type_id = self_type_id,
+                   .type_id = self_without_constraints_type_id,
                    .entity_name_id = entity_name_id,
                    // `None` because there is no equivalent non-symbolic value.
                    .value_id = SemIR::InstId::None,
                }));
-  auto existing =
-      context.scope_stack().LookupOrAddName(SemIR::NameId::PeriodSelf, inst_id);
+  auto existing = context.scope_stack().LookupOrAddName(
+      SemIR::NameId::PeriodSelf, period_self_inst_id);
   // Shouldn't have any names in newly created scope.
   CARBON_CHECK(!existing.has_value());
 
   // Save the `.Self` symbolic binding on the node stack. It will become the
   // first argument to the `WhereExpr` instruction.
-  context.node_stack().Push(node_id, inst_id);
+  context.node_stack().Push(node_id, period_self_inst_id);
 
   // Going to put each requirement on `args_type_info_stack`, so we can have an
   // inst block with the varying number of requirements but keeping other
   // instructions on the current inst block from the `inst_block_stack()`.
   context.args_type_info_stack().Push();
+
+  // Pass along all the constraints from the base facet type to be added to the
+  // resulting facet type.
+  context.args_type_info_stack().AddInstId(
+      AddInstInNoBlock<SemIR::RequirementBaseFacetType>(
+          context, SemIR::LocId(node_id),
+          {.base_type_inst_id =
+               context.types().GetInstId(self_with_constraints_type_id)}));
+
+  // Add a context stack for tracking rewrite constraints, that will be used to
+  // allow later constraints to read from them eagerly.
+  context.rewrites_stack().emplace_back();
```

Comment:
Do we need to add rewrites from `self_with_constraints_type_id` to the rewrite stack here?

Reply:
Yep! Test added, thanks.